flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/AffinityPropagation.java

```diff
@@ -0,0 +1,535 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.graph.library;
+
+import org.apache.flink.api.common.aggregators.LongSumAggregator;
+import org.apache.flink.api.common.functions.FilterFunction;
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.graph.EdgeDirection;
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.GraphAlgorithm;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.graph.Edge;
+import org.apache.flink.graph.EdgesFunction;
+import org.apache.flink.graph.spargel.MessageIterator;
+import org.apache.flink.graph.spargel.MessagingFunction;
+import org.apache.flink.graph.spargel.ScatterGatherConfiguration;
+import org.apache.flink.graph.spargel.VertexUpdateFunction;
+import org.apache.flink.types.LongValue;
+import org.apache.flink.types.NullValue;
+import org.apache.flink.util.Collector;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+
+
+/**
+ * This is an implementation of the Binary Affinity Propagation algorithm using a scatter-gather iteration.
+ * Note that is not the original Affinity Propagation.
+ *
+ * The input is an undirected graph where the vertices are the points to be clustered and the edge weights are the
+ * similarities of these points among them.
+ *
+ * The output is a Dataset of Tuple2, where f0 is the point id and f1 is the exemplar, so the clusters will be the
+ * the Tuples grouped by f1
+ *
+ * @see <a href="http://www.psi.toronto.edu/pubs2/2009/NC09%20-%20SimpleAP.pdf">
+ */
+
+@SuppressWarnings("serial")
+public class AffinityPropagation implements GraphAlgorithm<Long,NullValue,Double,DataSet<Tuple2<Long, Long>>> {
+
+	private static Integer maxIterations;
+	private static float damping;
+	private static float epsilon;
+
+	/**
+	 * Creates a new AffinityPropagation instance algorithm instance.
+	 *
+	 * @param maxIterations The maximum number of iterations to run
+	 * @param damping Damping factor.
+	 * @param epsilon Epsilon factor. Do not send message to a neighbor if the new message
+	 * has not changed more than epsilon.
+	 */
+	public AffinityPropagation(Integer maxIterations, float damping, float epsilon) {
+		this.maxIterations = maxIterations;
+		this.damping = damping;
+		this.epsilon = epsilon;
+	}
+
+	@Override
+	public DataSet<Tuple2<Long, Long>> run(Graph<Long, NullValue, Double> input) throws Exception {
+
+		// Create E and I AP vertices
+		DataSet<Vertex<Long, APVertexValue>> verticesWithAllInNeighbors =
+			input.groupReduceOnEdges(new InitAPVertex(), EdgeDirection.IN);
+
+		List<Vertex<Long, APVertexValue>> APvertices = verticesWithAllInNeighbors.collect();
```

Comment:
I would strongly suggest _against_ using `collect()` here. This will send all data back to the client and that's a potentially big dataset you're collecting.